README.md
# Sub-Address KEy (SAKE) filter

[![Build Status](https://github.com/breard-r/opensmtpd-filter-sake/actions/workflows/ci.yml/badge.svg)](https://github.com/breard-r/opensmtpd-filter-sake/actions/workflows/ci.yml)
![Minimal rustc version](https://img.shields.io/badge/rustc-1.70.0+-lightgray.svg)
![License MIT OR Apache 2.0](https://img.shields.io/badge/license-MIT%20OR%20Apache--2.0-blue)

Sub-address key filter for OpenSMTPD.


## What is the purpose of this project?

When giving your email address to someone, for example when registering an account on a website, a good practice is to give an unique address for each recipient. This way, if you start receiving spam on this unique address, you know which service has leaked your email address. One common way to do so is to use the sub-address delimiter (by default, the character `+`) in order to add a part with the service's name.

The problem is, people know that whatever is after the `+` can be discarded, and therefore some services drops it. This can also happen after a leak if the spammer doesn't want you to know which website has been breached. Furthermore, a spammer could also add a custom part after the `+` in order to cover its tracks.

Changing the default sub-address delimiter is a good idea, but isn't completely secure: in most cases, anyone will see the pattern you are using and will be able to deduce your email address for other services. For instance, if someone knows that you registered on `www.acme-corp.example.com` using the address `darra.acme-corp@mail.example.org` and on `www.super-social.example.com` using the address `darra.super-social@mail.example.org`, this person will deduce that your address on any service named `x` will be `darra.x@mail.example.org`.

This filter adds a way to configure some addresses (or aliases) in a way that the part after the sub-address delimiter includes a verification code that cannot be guessed. Following the previous example using the dot instead of the plus character as a sub-address delimiter, the addresses could be `darra.acme-corp.nbvtenby@mail.example.org` and `darra.super-social.heywkmrx@mail.example.org`. As you can see, a 5 bytes code in base32 has been added after the second delimiter.

This verification code is derived from both the address itself and a private key that only you and your email server know. Any email sent to this address without a valid authentication code will be rejected with `550 No such recipient here`. Therefore, one must know the private key in order to generate new valid addresses, which means only you can do so.

For more information on how to use it, please read the FAQ.


## Building

```
cargo build --release
```

The executable should be located in the `./target/release/` directory.


## Usage

The filter accepts the following options:

- `--address` or `-a`: specify an address where the filter will enforce the presence of a valid verification code (see below for the format)
- `--address-file` or `-A`: the path to a file where each line is an address as specified in `--address`
- `--separator` or `-s`: set the sub-address delimiter character (default: `+`) : this must match the character defined in `smtpd.conf` using `smtp sub-addr-delim`

An address must be composed of the following elements:
- the local part
- (optional) an `@` followed by a domain name
- the `:` character
- the private key in base64 (with padding)

Specifying a domain name configures the filter to match addresses on both the local part and the specified domain name. If no domain name is specified, the match will be on the local part only, and therefore all domain names will be accepted.

The `--address` option may be specified multiples times and can also be combined with `--address-file`.

In an address file, empty lines and lines starting with the `#` character are ignored.

The private key's length must be either 128 or 256 bits. To generate a 128 bits key, the following command is recommended:

```
openssl rand -base64 16
```

Example configuration:

```
# Sub-addresses
smtp sub-addr-delim "+"
filter "sake" proc-exec "filter-sake -s '+' -a 'a@example.org:11voiefK5PgCX5F1TTcuoQ==' -a 'b:3pUdigGQNXYBeKJdYDdERQ=='"

# Tables
table domains { "example.org", "example.com" }
table vusers { "test" = "1000:100:/var/vmail/test", "b" = "1000:100:/var/vmail/b" }
table aliases { "a" = "test" }

# Listening
listen on 127.0.0.1 hostname localhost filter "sake"
listen on ::1 hostname localhost filter "sake"

# Delivering
action "deliver" maildir userbase <vusers> alias <aliases>
match from any for domain <domains> action "deliver"
```


## Code generation protocol

Let start with some definition. For this protocol, an email address is composed of a local part, a sub-address delimiter, a sub-address, another sub-address delimiter, the validation code, the at sign and the domain name. For instance, for the address `darra.service.gizti5lj@mail.example.org`:
- local part: `darra`
- sub-address delimiter: `.`
- sub-address: `service`
- validation code: `gizti5lj`
- domain name: `mail.example.org`

The code generation protocol is based on the HMAC-SHA-256 function. The hasher is configured with the private key, then the following data is hashed, in this order: the local part, the sub-address delimiter and the sub-address.

This hash is then reduced to 5 bytes using the following dynamic offset truncation method. From the last byte of the hash, we take the last 4 bits, which gives an offset between 0 and 15. We then take the 5 bytes of the hash located at this offset.

The code is then generated by encoding those 5 bytes using base32 ([RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648)) without padding.


## Frequently Asked Questions

### How do I generate valid addresses?

The filter itself is useful for OpenSMTPD only, it is not meant to be used directly by the user. For this usage, you should use [sake-app](https://github.com/breard-r/sake-app).

In the event you do not wish to host it yourself, you can use [https://sake.email/](https://sake.email/). This is a client-side only application, your data will be exclusively stored in your browser's local storage.

### Do I need to have several mailboxes?

No.

The local part can either be a real mailbox or an alias. It is up to you to decide how to setup you mail server.

### Does it works with Postfix / Exim / whatever?

No, this project is based on the filter API used by OpenSMTPD.

### Does it supports IDN?

Yes, internationalized domain names (IDN) are supported. You can specify domain names either using valid UTF-8 or Punycode ([RFC 3492](https://datatracker.ietf.org/doc/html/rfc3492)).

### How long should be my private key?

Privates keys must have a length of either 128 bits (16 bytes) or 256 bits (32 bytes). Unless you have some very specific needs, you should choose a 128 bits key.

### What about key rotation?

Rotating the key would mean that all previously generated addresses for this local part would suddenly be invalid. Therefore, the key associated with a local part must not change.

That said, you can add a new local part that uses a new key and stop using the previous one. To this end, it is recommended to use discardable names. Local parts composed of one to three characters without special meaning are good candidates to this.

### Is the code cryptographically secure?

No, it is not.

Efforts have been made so it is almost impossible to use one or several known valid addresses to create new addresses or recover the key. However, it may not be considered cryptographically secure because of code's short length (5 bytes).

### Can you detail the efforts made to get a mostly secure code?

At the time of writing, the HMAC-SHA-256 function has no known vulnerability.

The dynamic offset truncation is a simplified version of the one defined in [section 5.3 of RFC 4226](https://datatracker.ietf.org/doc/html/rfc4226#section-5.3) (HOTP: An HMAC-Based One-Time Password Algorithm). However, reducing the 32 bytes output into 5 bytes can never be considered completely secure.

Therefore, although it cannot be considered cryptographically secure, efforts have been made to generate a code that is sufficiently resistant to most attackers.

If your threat-model includes attackers that are backed by a government or a powerful criminal organization, you should seek for professional help instead of trusting random projects on the internet.


CHANGELOG.md
[//]: # (Copyright 2023 Rodolphe Bréard <v.code.ngyvz7vf@what.tf>)

[//]: # (Copying and distribution of this file, with or without modification,)
[//]: # (are permitted in any medium without royalty provided the copyright)
[//]: # (notice and this notice are preserved.  This file is offered as-is,)
[//]: # (without any warranty.)

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2023-08-25

### Added

- Initial release.


start_test.py
#!/usr/bin/env python3

import os
import pathlib
import shutil
import smtplib
import stat
import subprocess
import sys
import tempfile
import time

ADDR_FROM = "test.from@example.org"
DEFAULT_PORT = 2525


def fail(message):
    print(message, file=sys.stderr)
    sys.exit(1)


def cp_tmp_file(path, executable=False):
    file = tempfile.NamedTemporaryFile(suffix=f"-{path.name}", delete=False)
    with open(path, mode="rb") as f:
        file.write(f.read())
        file.flush()
    flags = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IROTH
    if executable:
        flags = flags | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH
    os.chmod(file.name, flags)
    file.close()
    return file


def get_filter_path(test_dir, target):
    filter_path = test_dir / "target" / target / "filter-sake"
    filter_path = cp_tmp_file(filter_path, executable=True).name
    return filter_path


def get_opensmtpd_config(port, filter_exe, maildir_path):
    cfg_content = f"""# OpenSMTPD test configuration

# Sub-addresses
filter "sake" proc-exec "{filter_exe} --address 'a@example.org:11voiefK5PgCX5F1TTcuoQ==' --address 'b:11voiefK5PgCX5F1TTcuoQ=='"

# Tables
table domains {{ "example.org", "example.com" }}
table vusers {{ "test" = "1000:100:{maildir_path}", "b" = "1000:100:{maildir_path}" }}
table aliases {{ "a" = "test" }}

# Listening
listen on 127.0.0.1 port {port} hostname localhost filter "sake"
listen on ::1 port {port} hostname localhost filter "sake"

# Delivering
action "deliver" maildir userbase <vusers> alias <aliases>
match from any for domain <domains> action "deliver"
"""
    cfg_file = tempfile.NamedTemporaryFile(prefix="smtpd-", suffix=".conf")
    cfg_file.write(cfg_content.encode())
    cfg_file.flush()
    return cfg_file


def send_msg(smtp, to_addr, is_valid):
    msg = f"From: {ADDR_FROM}\r\nTo: {to_addr}\r\n\r\ntest"
    try:
        smtp.sendmail(ADDR_FROM, to_addr, msg)
        if not is_valid:
            print(f"{to_addr}: accepted: error")
            return False
        print(f"{to_addr}: accepted: ok")
    except smtplib.SMTPRecipientsRefused:
        if is_valid:
            print(f"{to_addr}: refused: error")
            return False
        print(f"{to_addr}: refused: ok")
    return True


def start_opensmtpd(cfg_path):
    args = [
        shutil.which("sudo"),
        shutil.which("smtpd"),
        "-d",
        "-f",
        cfg_path.name,
    ]
    p = subprocess.Popen(
        args,
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    time.sleep(8)
    return p.pid


def kill_opensmtpd(pid):
    if pid is not None:
        subprocess.Popen([shutil.which("sudo"), shutil.which("kill"), f"{pid}"])


def get_maildir():
    maildir = tempfile.TemporaryDirectory(prefix="Maildir_")
    flags = (
        stat.S_IRUSR
        | stat.S_IWUSR
        | stat.S_IXUSR
        | stat.S_IRGRP
        | stat.S_IWGRP
        | stat.S_IXGRP
        | stat.S_IROTH
        | stat.S_IWOTH
        | stat.S_IXOTH
    )
    os.chmod(maildir.name, flags)
    return maildir


def start_tests(test_dir, smtp_port):
    to_addrs = [
        ("test@example.org", True),
        ("a@example.com", True),
        ("a+anything@example.com", True),
        ("a+anything+test@example.com", True),
        ("a+test+pxmnoheo@example.org", True),
        ("b+test+uivojtoa@example.org", True),
        ("test@nope.example.org", False),
        ("a@example.org", False),
        ("a+invalid@example.org", False),
        ("a+invalid+input@example.org", False),
        ("b@example.org", False),
    ]
    maildir = get_maildir()
    filter_cmd = get_filter_path(test_dir, "debug")
    pid_smtpd = None
    has_errors = False
    try:
        cfg_path = get_opensmtpd_config(smtp_port, filter_cmd, maildir.name)
        pid_smtpd = start_opensmtpd(cfg_path)
        with smtplib.SMTP(host="localhost", port=smtp_port) as smtp_session:
            for addr, is_valid in to_addrs:
                if not send_msg(smtp_session, addr, is_valid):
                    has_errors = True
    except Exception:
        kill_opensmtpd(pid_smtpd)
        raise
    kill_opensmtpd(pid_smtpd)
    if has_errors:
        fail("test failed")


def main():
    test_dir = pathlib.Path(__file__).parent.resolve()
    os.chdir(test_dir.parent)
    start_tests(test_dir, DEFAULT_PORT)


if __name__ == "__main__":
    main()


pwa/sw.js
const sw_version = '0.1.0';
const cache_name = `sake-v${sw_version}`;
const cached_files = [
	'/',
	'/app.js',
	'/index.html',
	'/vendor/base32.min.js',
	'/vendor/base32.min.js.map',
	'/vendor/sha256.min.js',
];

function log_message(msg) {
	console.log(`[Service Worker] v${sw_version}: ${msg}`);
}

self.addEventListener('install', (event) => {
	log_message('Installed');
	self.skipWaiting();
	event.waitUntil(caches.open(cache_name).then((cache) => {
		log_message('Caching all');
		return cache.addAll(cached_files);
	}));
});

self.addEventListener('activate', (e) => {
	e.waitUntil(
		caches.keys().then((keyList) =>
			Promise.all(
				keyList.map((key) => {
					if (key != cache_name) {
						log_message(`Cleaning cache: ${key}`);
						return caches.delete(key);
					}
				}),
			),
		),
	);
	e.waitUntil(clients.claim());
	log_message('Active');
});

self.addEventListener('fetch', (event) => {
	if (!(event.request.url.startsWith('https:') || event.request.url.startsWith('http:'))) {
		log_message(`Fetching resource failed: invalid protocol: ${event.request.url}`);
		return;
	}

	event.respondWith(caches.open(cache_name).then((cache) => {
		log_message(`Fetching resource: ${event.request.url}`);
		return cache.match(event.request).then((cached_response) => {
			const fetched_response = fetch(event.request).then((network_response) => {
				log_message(`Caching resource: ${event.request.url}`);
				cache.put(event.request, network_response.clone());
				return network_response;
			});
			return cached_response || fetched_response;
		});
	}));
});


pwa/app.js
function base32_nopad_encode(slice) {
	const encoder = new base32.Encoder({ type: "rfc4648", lc: true });
	const code = encoder.write(slice).finalize();
	return code.replaceAll('=', '');
}

function base64_decode(str_b64) {
	const raw_str = atob(str_b64);
	const length = raw_str.length;
	var b = [];
	for (var i = 0; i < length; i++) {
		b.push(raw_str.charCodeAt(i));
	}
	return b;
}

class AccountValueError extends Error {
	constructor(field_id, ...params) {
		super(...params);
		if (Error.captureStackTrace) {
			Error.captureStackTrace(this, AccountValueError);
		}
		this.name = "AccountValueError";
		this.field_id = field_id;
	}
}

class Account {
	constructor(local_part, separator, domain, key) {
		// Set the local part
		if (!local_part) {
			throw new AccountValueError("new-addr-local-part", "The local part cannot be empty.");
		}
		this.local_part = local_part;

		// Set the separator
		if (!separator || separator.length !== 1) {
			throw new AccountValueError("new-addr-separator", "The separator must be a single character.");
		}
		this.separator = separator;

		// Set the domain name
		if (!domain) {
			throw new AccountValueError("new-addr-domain", "The domain cannot be empty.");
		}
		this.domain = domain;

		// Set the private key
		if (Array.isArray(key)) {
			this.key = key;
		} else {
			try {
				this.key = base64_decode(key);
			} catch (e) {
				console.log(e);
				throw new AccountValueError("new-addr-key", "The key must be a valid base64 string.");
			}
			if (!this.key.length) {
				throw new AccountValueError("new-addr-key", "The key must not be empty.");
			}
		}
	}

	getName() {
		return `${this.local_part}@${this.domain}`;
	}

	genSubAddr(sub_addr_name) {
		var hasher = sha256.hmac.create(this.key);
		hasher.update(this.local_part);
		hasher.update(this.separator);
		hasher.update(sub_addr_name);
		const hash = hasher.array();
		const offset = hash[hash.length - 1] & 0xf;
		const reduced_hash = hash.slice(offset, offset + 5);
		const code = base32_nopad_encode(reduced_hash);
		return `${this.local_part}${this.separator}${sub_addr_name}${this.separator}${code}@${this.domain}`
	}

	register() {
		// Handle duplicate
		if (localStorage.getItem(this.getName())) {
			throw new AccountValueError("", "The account already exists.");
		}
		localStorage.setItem(this.getName(), JSON.stringify(this));
	}
}

document.addEventListener('DOMContentLoaded', () => {
	// Functions to display and remove the error message
	function setErrorMessage(message) {
		unsetErrorMessage();
		const el = document.createElement('p');
		el.classList.add('notification');
		el.classList.add('is-danger');
		el.classList.add('is-light');
		el.appendChild(document.createTextNode(message));
		document.querySelector('#new-account-error-msg-container').appendChild(el);
	}

	function unsetErrorMessage() {
		const el_cont = document.querySelector('#new-account-error-msg-container');
		while (el_cont.lastElementChild) {
			el_cont.removeChild(el_cont.lastElementChild);
		}
		['#new-addr-local-part', '#new-addr-separator', '#new-addr-domain', '#new-addr-key'].forEach((selector) => {
			document.querySelector(selector).classList.remove('is-danger');
		});
	}

	// Functions to open and close a modal
	function openModal(el) {
		if (el.id === 'modal-del-account') {
			const name_el = document.createTextNode(getSelectedAccountName());
			document.querySelector('#del-account-name').appendChild(name_el);
		}
		el.classList.add('is-active');
	}

	function openNewAccountModal(el) {
		const new_account_modal = document.querySelector('#modal-add-account');
		openModal(new_account_modal);
	}

	function closeModal(el) {
		if (!(el.id === 'modal-add-account' && localStorage.length === 0)) {
			el.classList.remove('is-active');
		}
	}

	function closeAllModals() {
		(document.querySelectorAll('.modal') || []).forEach((modal) => {
			closeModal(modal);
		});
	}

	// Function to get the name of the currently selected account
	function getSelectedAccountName() {
		return document.querySelector('#account-name').value;
	}

	// Function to get an account by its name
	function getAccountByName(name) {
		const account_string = localStorage.getItem(name);
		if (!account_string) {
			return null;
		}
		const account_raw = JSON.parse(account_string);
		return new Account(
			account_raw.local_part,
			account_raw.separator,
			account_raw.domain,
			account_raw.key,
		);
	}

	// Function to synchronize the account list
	function syncAccountList() {
		var acc_list = document.querySelector('#account-name');
		while (acc_list.lastElementChild) {
			acc_list.removeChild(acc_list.lastElementChild);
		}
		var account_names = [];
		for (var i = 0, len = localStorage.length; i < len; ++i) {
			account_names.push(localStorage.key(i));
		}
		account_names.sort();
		for (const name of account_names) {
			const account = getAccountByName(name);
			const new_elem = new Option(account.getName(), account.getName());
			acc_list.appendChild(new_elem);
		}
		if (localStorage.length === 0) {
			openNewAccountModal();
		}
	}
	syncAccountList();

	// Add a click event on buttons to open a specific modal
	(document.querySelectorAll('.js-modal-trigger') || []).forEach((trigger) => {
		const modal = trigger.dataset.target;
		const target = document.getElementById(modal);

		trigger.addEventListener('click', () => {
			openModal(target);
		});
	});

	// Add a click event on various child elements to close the parent modal
	(document.querySelectorAll('.modal-background, .modal-close, .modal-card-head .delete, .modal-card-foot .button-close') || []).forEach((close) => {
		const target = close.closest('.modal');

		close.addEventListener('click', () => {
			closeModal(target);
		});
	});

	// Add a keyboard event to close all modals
	document.addEventListener('keydown', (event) => {
		if (event.code === 'Escape') {
			closeAllModals();
		}
	});

	// Add a click event on the delete account button to display the confirmation message
	document.querySelector('#btn-del-account-confirm').addEventListener('click', (event) => {
		const account_name = getSelectedAccountName();
		localStorage.removeItem(account_name);
		console.log(`Account ${account_name} deleted.`);
		syncAccountList();
		closeAllModals();
	});

	// Add a click event on the new account button to register the new account
	document.querySelector('#btn-new-account').addEventListener('click', (event) => {
		console.log('Adding new account…');
		try {
			const new_account = new Account(
				document.querySelector('#new-addr-local-part').value,
				document.querySelector('#new-addr-separator').value,
				document.querySelector('#new-addr-domain').value,
				document.querySelector('#new-addr-key').value,
			);
			console.log(new_account);
			new_account.register();
			console.log(`Account ${new_account.getName()} added.`);
			['#new-addr-local-part', '#new-addr-domain', '#new-addr-key'].forEach((selector) => {
				document.querySelector(selector).value = '';
			});
			document.querySelector('#new-addr-separator').value = '+';
			syncAccountList();
			closeAllModals();
		} catch (e) {
			console.log(`${e.name}: ${e.field_id}: ${e.message}`);
			setErrorMessage(e.message);
			if (e.field_id) {
				document.getElementById(e.field_id).classList.add('is-danger');
			}
		}
	});

	// Add a click event on the new address button to generate it
	document.querySelector('#btn-generate').addEventListener('click', (event) => {
		event.preventDefault();
		const account = getAccountByName(document.querySelector('#account-name').value);
		const new_address = account.genSubAddr(document.querySelector('#sub-addr-name').value);
		document.querySelector('#generated-addr').value = new_address;
		console.log(`New sub-address for account ${account.getName()}: ${new_address}`);
	});

	// Add a change event on the main form to remove previously generated address
	['#account-name', '#sub-addr-name'].forEach((selector) => {
		document.querySelector(selector).addEventListener('change', () => {
			document.querySelector('#generated-addr').value = '';
		});
	});
});

window.addEventListener('load', () => {
	if ('serviceWorker' in navigator) {
		navigator.serviceWorker.register('/sw.js');
	}
});


src/code.rs
use hmac::{Hmac, Mac};
use sha2::Sha256;

const NB_BYTES: usize = 5;

pub fn generate_code(local_part: &str, separator: char, sub_addr: &str, key: &[u8]) -> Vec<u8> {
	// Compute the HMAC-SHA-256
	let mut hmac = Hmac::<Sha256>::new_from_slice(key).unwrap();
	hmac.update(local_part.as_bytes());
	hmac.update(separator.to_string().as_bytes());
	hmac.update(sub_addr.as_bytes());
	let result = hmac.finalize().into_bytes();

	// Reduce the result to NB_BYTES using a dynamic offset truncation
	let offset = (result[result.len() - 1] & 0xf) as usize;
	result[offset..offset + NB_BYTES].to_vec()
}

#[cfg(test)]
mod tests {
	use super::generate_code;

	#[test]
	fn code_generation() {
		let key: &[u8] = &[
			0xd7, 0x5b, 0xe8, 0x89, 0xe7, 0xca, 0xe4, 0xf8, 0x02, 0x5f, 0x91, 0x75, 0x4d, 0x37,
			0x2e, 0xa1,
		];
		let code = generate_code("a", '+', "test", key);
		assert_eq!(code, vec![0x7d, 0xd8, 0xd7, 0x1c, 0x8e]);
	}
}


src/input.rs
use crate::address::CodedAddress;
use anyhow::{anyhow, ensure, Result};

macro_rules! next_param {
	($it: ident) => {
		$it.next().ok_or(anyhow!("missing parameter"))
	};
}

#[derive(Debug)]
pub struct Input {
	session_id: String,
	token: String,
	address: CodedAddress,
}

impl Input {
	pub fn answer(&self, msg: &str) {
		println!("filter-result|{0}|{1}|{msg}", self.session_id, self.token);
	}

	pub fn get_coded_address(&self) -> &CodedAddress {
		&self.address
	}
}

pub fn parse_input(input: &str, separator: char) -> Result<Input> {
	let mut params_it = input.split(crate::PARAM_SEPARATOR);
	let stream = next_param!(params_it)?;
	ensure!(stream == "filter", "{stream}: invalid stream");
	let version = next_param!(params_it)?;
	ensure!(!version.is_empty(), "empty version");
	let timestamp = next_param!(params_it)?;
	ensure!(!timestamp.is_empty(), "empty timestamp");
	let subsystem = next_param!(params_it)?;
	ensure!(subsystem == "smtp-in", "{subsystem}: invalid subsystem");
	let filter = next_param!(params_it)?;
	ensure!(filter == "rcpt-to", "{filter}: invalid filter");
	let session_id = next_param!(params_it)?.to_string();
	ensure!(!session_id.is_empty(), "empty session id");
	let token = next_param!(params_it)?.to_string();
	ensure!(!token.is_empty(), "empty token");
	let address = next_param!(params_it)?.trim_end();
	let address = CodedAddress::parse(address, separator)?;
	Ok(Input {
		session_id,
		token,
		address,
	})
}

#[cfg(test)]
mod tests {
	use super::parse_input;

	#[test]
	fn test_valid_input() {
		let input = "filter|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		let res = parse_input(input, '+');
		assert!(res.is_ok());
		let inp = res.unwrap();
		assert_eq!(inp.session_id, "7641df9771b4ed00");
		assert_eq!(inp.token, "1ef1c203cc576e5d");
	}

	#[test]
	fn test_empty_input() {
		assert!(parse_input("", '+').is_err());
	}

	#[test]
	fn test_empty_stream() {
		let input = "|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_invalid_stream() {
		let input = "invalid|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_version() {
		let input = "filter||1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_timestamp() {
		let input =
			"filter|0.5||smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_subsystem() {
		let input = "filter|0.5|1576146008.006099||rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_invalid_subsystem() {
		let input = "filter|0.5|1576146008.006099|invalid|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_phase() {
		let input = "filter|0.5|1576146008.006099|smtp-in||7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_invalid_phase() {
		let input = "filter|0.5|1576146008.006099|smtp-in|invalid|7641df9771b4ed00|1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_id() {
		let input =
			"filter|0.5|1576146008.006099|smtp-in|rcpt-to||1ef1c203cc576e5d|derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_token() {
		let input =
			"filter|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00||derp@example.com";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_empty_data() {
		let input =
			"filter|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|";
		assert!(parse_input(input, '+').is_err());
	}

	#[test]
	fn test_missing_data() {
		let input =
			"filter|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d";
		assert!(parse_input(input, '+').is_err());
	}
}


src/config.rs
use crate::address::KeyedAddress;
use anyhow::Result;
use clap::Parser;
use std::collections::HashSet;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::PathBuf;
use std::str::FromStr;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
pub struct Config {
	#[arg(short, long)]
	address: Vec<String>,
	#[arg(short = 'A', long)]
	address_file: Option<PathBuf>,
	#[arg(short, long, default_value_t = crate::DEFAULT_SEPARATOR)]
	separator: char,
}

impl Config {
	pub fn addresses(&self) -> Result<HashSet<KeyedAddress>> {
		let mut addr_set = HashSet::new();
		for addr in &self.address {
			addr_set.insert(KeyedAddress::from_str(addr)?);
		}
		if let Some(path) = &self.address_file {
			let f = File::open(path)?;
			let f = BufReader::new(f);
			for line in f.lines() {
				let line = line?;
				let addr = line.trim();
				if !addr.is_empty() && !addr.starts_with(crate::COMMENT_CHAR) {
					addr_set.insert(KeyedAddress::from_str(addr)?);
				}
			}
		}
		Ok(addr_set)
	}

	pub fn get_separator(&self) -> char {
		self.separator
	}
}


src/service.rs
use crate::address::KeyedAddress;
use crate::config::Config;
use crate::input::{parse_input, Input};
use anyhow::{anyhow, Result};
use clap::Parser;
use std::collections::HashSet;
use std::io;

pub const CONFIG_END: &str = "config|ready";
pub const ANSWER_OK: &str = "proceed";
pub const ANSWER_ERR: &str = "reject|550 No such recipient here";

macro_rules! next_line {
	($it: ident) => {
		$it.next()
			.ok_or(anyhow!("standard input has been closed"))??
	};
}

pub fn start_service() -> Result<()> {
	let cfg = Config::parse();
	let addresses = cfg.addresses()?;
	let mut lines = io::stdin().lines();

	// Handshake
	loop {
		let input = next_line!(lines);
		if input.trim_end() == CONFIG_END {
			break;
		}
	}
	println!("register|filter|smtp-in|rcpt-to");
	println!("register|ready");

	// Input processing
	loop {
		let input = next_line!(lines);
		if !input.is_empty() {
			match parse_input(&input, cfg.get_separator()) {
				Ok(input) => {
					if allow_email(&input, &addresses) {
						input.answer(ANSWER_OK);
					} else {
						input.answer(ANSWER_ERR);
					}
				}
				Err(e) => {
					eprintln!("error: {e:#}");
				}
			}
		}
	}
}

fn allow_email(input: &Input, addr_lst: &HashSet<KeyedAddress>) -> bool {
	let address = input.get_coded_address();
	for addr_k in addr_lst {
		if addr_k == address {
			return addr_k.check_code(address);
		}
	}
	true
}

#[cfg(test)]
mod tests {
	use super::allow_email;
	use crate::input::parse_input;
	use crate::service::KeyedAddress;
	use std::collections::HashSet;
	use std::str::FromStr;

	fn run_test_with_addr(address: &str) -> bool {
		// Preparing the input
		let input_str = format!("filter|0.5|1576146008.006099|smtp-in|rcpt-to|7641df9771b4ed00|1ef1c203cc576e5d|{address}");
		let input = parse_input(&input_str, '+');
		assert!(input.is_ok());
		let input = input.unwrap();

		// Preparing the test keyed addresses
		let kaddr_lst = [
			"a@example.org:11voiefK5PgCX5F1TTcuoQ==",
			"b:11voiefK5PgCX5F1TTcuoQ==",
		];
		let mut addr_set = HashSet::with_capacity(kaddr_lst.len());
		for kaddr_str in kaddr_lst {
			let keyed_addr = KeyedAddress::from_str(kaddr_str);
			assert!(keyed_addr.is_ok());
			let keyed_addr = keyed_addr.unwrap();
			addr_set.insert(keyed_addr);
		}

		// Run the test
		allow_email(&input, &addr_set)
	}

	#[test]
	fn test_valid_code_domain() {
		assert!(run_test_with_addr("a+test+pxmnoheo@example.org"));
	}

	#[test]
	fn test_valid_code_no_domain() {
		assert!(run_test_with_addr("b+test+uivojtoa@example.org"));
	}

	#[test]
	fn test_invalid_code_domain() {
		assert!(!run_test_with_addr("a+test+orsxg5a@example.org"));
	}

	#[test]
	fn test_invalid_code_no_domain() {
		assert!(!run_test_with_addr("b+test+orsxg5a@example.org"));
	}

	#[test]
	fn test_no_code() {
		assert!(!run_test_with_addr("a+test@example.org"));
	}

	#[test]
	fn test_no_sub_addr() {
		assert!(!run_test_with_addr("b@example.org"));
	}

	#[test]
	fn test_different_domain() {
		assert!(run_test_with_addr("a@example.com"));
	}
}
